File: nRF Toolbox/BPM/BPMViewController.m
/*
 * Copyright (c) 2015, Nordic Semiconductor
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the
 * documentation and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote products derived from this
 * software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
 * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 * USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#import "BPMViewController.h"
#import "ScannerViewController.h"
#import "Constants.h"
#import "AppUtilities.h"
#import "CharacteristicReader.h"

@interface BPMViewController () {
    CBUUID *bpmServiceUUID;
    CBUUID *bpmBloodPressureMeasurementCharacteristicUUID;
    CBUUID *bpmIntermediateCuffPressureCharacteristicUUID;
    CBUUID *batteryServiceUUID;
    CBUUID *batteryLevelCharacteristicUUID;
}

/*!
 * This property is set when the device successfully connects to the peripheral. It is used to cancel the connection
 * after user press Disconnect button.
 */
@property (strong, nonatomic) CBPeripheral *connectedPeripheral;

@property (weak, nonatomic) IBOutlet UILabel *systolic;
@property (weak, nonatomic) IBOutlet UILabel *systolicUnit;
@property (weak, nonatomic) IBOutlet UILabel *diastolic;
@property (weak, nonatomic) IBOutlet UILabel *diastolicUnit;
@property (weak, nonatomic) IBOutlet UILabel *meanAp;
@property (weak, nonatomic) IBOutlet UILabel *meanApUnit;
@property (weak, nonatomic) IBOutlet UILabel *pulse;
@property (weak, nonatomic) IBOutlet UILabel *timestamp;

- (IBAction)aboutButtonClicked:(id)sender;

@end

@implementation BPMViewController
@synthesize bluetoothManager;
@synthesize verticalLabel;
@synthesize battery;
@synthesize deviceName;
@synthesize connectButton;
@synthesize connectedPeripheral;


-(id)initWithCoder:(NSCoder *)aDecoder
{
    self = [super initWithCoder:aDecoder];
    if (self) {
        // Custom initialization
        bpmServiceUUID = [CBUUID UUIDWithString:bpmServiceUUIDString];
        bpmBloodPressureMeasurementCharacteristicUUID = [CBUUID UUIDWithString:bpmBloodPressureMeasurementCharacteristicUUIDString];
        bpmIntermediateCuffPressureCharacteristicUUID = [CBUUID UUIDWithString:bpmIntermediateCuffPressureCharacteristicUUIDString];
        batteryServiceUUID = [CBUUID UUIDWithString:batteryServiceUUIDString];
        batteryLevelCharacteristicUUID = [CBUUID UUIDWithString:batteryLevelCharacteristicUUIDString];
    }
    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    // Rotate the vertical label
    self.verticalLabel.transform = CGAffineTransformRotate(CGAffineTransformMakeTranslation(-150.0f, 0.0f), (float)(-M_PI / 2));
}

-(void)appDidEnterBackground:(NSNotification *)_notification
{
    [AppUtilities showBackgroundNotification:[NSString stringWithFormat:@"You are still connected to %@ peripheral. It will collect data also in background.",connectedPeripheral.name]];
}

-(void)appDidBecomeActiveBackground:(NSNotification *)_notification
{
    [[UIApplication sharedApplication] cancelAllLocalNotifications];
}

- (IBAction)aboutButtonClicked:(id)sender {
    [self showAbout:[AppUtilities getBPMHelpText]];
}

- (IBAction)connectOrDisconnectClicked {
    if (connectedPeripheral != nil)
    {
        [bluetoothManager cancelPeripheralConnection:connectedPeripheral];
    }
}

-(BOOL)shouldPerformSegueWithIdentifier:(NSString *)identifier sender:(id)sender
{
    // The 'scan' seque will be performed only if connectedPeripheral == nil (if we are not connected already).
    return ![identifier isEqualToString:@"scan"] || connectedPeripheral == nil;
}

-(void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    if ([segue.identifier isEqualToString:@"scan"])
    {
        // Set this contoller as scanner delegate
        UINavigationController *nc = segue.destinationViewController;
        ScannerViewController *controller = (ScannerViewController *)nc.childViewControllerForStatusBarHidden;
        controller.filterUUID = bpmServiceUUID;
        controller.delegate = self;
    }
}

#pragma mark Scanner Delegate methods

-(void)centralManager:(CBCentralManager *)manager didPeripheralSelected:(CBPeripheral *)peripheral
{
    // Some devices disconnects just after finishing measurement so we have to clear the UI before new connection, not after previous.
    [self clearUI];
    
    // We may not use more than one Central Manager instance. Let's just take the one returned from Scanner View Controller
    bluetoothManager = manager;
    bluetoothManager.delegate = self;
    
    // The sensor has been selected, connect to it
    peripheral.delegate = self;
    NSDictionary *options = [NSDictionary dictionaryWithObject:[NSNumber numberWithBool:YES] forKey:CBConnectPeripheralOptionNotifyOnNotificationKey];
    [bluetoothManager connectPeripheral:peripheral options:options];
}

#pragma mark Central Manager delegate methods

- (void)centralManagerDidUpdateState:(CBCentralManager *)central
{
    if (central.state == CBCentralManagerStatePoweredOn) {
        // TODO
    }
    else
    {
        // TODO
        NSLog(@"Bluetooth not ON");
    }
}

- (void)centralManager:(CBCentralManager *)central didConnectPeripheral:(CBPeripheral *)peripheral
{
    // Scanner uses other queue to send events. We must edit UI in the main queue
    dispatch_async(dispatch_get_main_queue(), ^{
        [deviceName setText:peripheral.name];
        [connectButton setTitle:@"DISCONNECT" forState:UIControlStateNormal];
    });
    //Following if condition display user permission alert for background notification
    if ([UIApplication instancesRespondToSelector:@selector(registerUserNotificationSettings:)]) {
        [[UIApplication sharedApplication] registerUserNotificationSettings:[UIUserNotificationSettings settingsForTypes:UIUserNotificationTypeAlert|UIUserNotificationTypeSound categories:nil]];
    }
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(appDidEnterBackground:) name:UIApplicationDidEnterBackgroundNotification object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(appDidBecomeActiveBackground:) name:UIApplicationDidBecomeActiveNotification object:nil];
    
    // Peripheral has connected. Discover required services
    connectedPeripheral = peripheral;
    [peripheral discoverServices:@[bpmServiceUUID, batteryServiceUUID]];
}

-(void)centralManager:(CBCentralManager *)central didFailToConnectPeripheral:(CBPeripheral *)peripheral error:(NSError *)error
{
    // Scanner uses other queue to send events. We must edit UI in the main queue
    dispatch_async(dispatch_get_main_queue(), ^{
        [AppUtilities showAlert:@"Error" alertMessage:@"Connecting to the peripheral failed. Try again"];
        [connectButton setTitle:@"CONNECT" forState:UIControlStateNormal];
        connectedPeripheral = nil;
        
        [self clearUI];
    });
}

- (void)centralManager:(CBCentralManager *)central didDisconnectPeripheral:(CBPeripheral *)peripheral error:(NSError *)error
{
    // Scanner uses other queue to send events. We must edit UI in the main queue
    dispatch_async(dispatch_get_main_queue(), ^{
        [connectButton setTitle:@"CONNECT" forState:UIControlStateNormal];
        if ([AppUtilities isApplicationStateInactiveORBackground]) {
            [AppUtilities showBackgroundNotification:[NSString stringWithFormat:@"%@ peripheral is disconnected",peripheral.name]];
        }
        connectedPeripheral = nil;
        
        [[NSNotificationCenter defaultCenter] removeObserver:self name:UIApplicationDidBecomeActiveNotification object:nil];
        [[NSNotificationCenter defaultCenter] removeObserver:self name:UIApplicationDidEnterBackgroundNotification object:nil];
    });
}

- (void) clearUI
{
    [deviceName setText:@"DEFAULT BPM"];
    battery.tag = 0;
    [battery setTitle:@"n/a" forState:UIControlStateDisabled];
    
    self.systolicUnit.hidden = YES;
    self.diastolicUnit.hidden = YES;
    self.meanApUnit.hidden = YES;
    
    self.systolic.text = @"-";
    self.diastolic.text = @"-";
    self.meanAp.text = @"-";
    self.pulse.text = @"-";
    self.timestamp.text = @"n/a";
}

#pragma mark Peripheral delegate methods

-(void)peripheral:(CBPeripheral *)peripheral didDiscoverServices:(NSError *)error
{
    if (error)
    {
        NSLog(@"Error discovering service: %@", [error localizedDescription]);
        [bluetoothManager cancelPeripheralConnection:connectedPeripheral];
        return;
    }
    
    for (CBService *service in peripheral.services)
    {
        // Discovers the characteristics for a given service
        if ([service.UUID isEqual:bpmServiceUUID])
        {
            [connectedPeripheral discoverCharacteristics:@[bpmBloodPressureMeasurementCharacteristicUUID, bpmIntermediateCuffPressureCharacteristicUUID] forService:service];
        }
        else if ([service.UUID isEqual:batteryServiceUUID])
        {
            [connectedPeripheral discoverCharacteristics:@[batteryLevelCharacteristicUUID] forService:service];
        }
    }
}

-(void)peripheral:(CBPeripheral *)peripheral didDiscoverCharacteristicsForService:(CBService *)service error:(NSError *)error
{
    // Characteristics for one of those services has been found
    if ([service.UUID isEqual:bpmServiceUUID])
    {
        for (CBCharacteristic *characteristic in service.characteristics)
        {
            if ([characteristic.UUID isEqual:bpmBloodPressureMeasurementCharacteristicUUID] ||
                [characteristic.UUID isEqual:bpmIntermediateCuffPressureCharacteristicUUID])
            {
                // Enable notifications and indications on data characteristics
                [peripheral setNotifyValue:YES forCharacteristic:characteristic];
            }
        }
    }
    else if ([service.UUID isEqual:batteryServiceUUID])
    {
        for (CBCharacteristic *characteristic in service.characteristics)
        {
            if ([characteristic.UUID isEqual:batteryLevelCharacteristicUUID])
            {
                // Read the current battery value
                [peripheral readValueForCharacteristic:characteristic];
                break;
            }
        }
    }
}

-(void)peripheral:(CBPeripheral *)peripheral didUpdateValueForCharacteristic:(CBCharacteristic *)characteristic error:(NSError *)error
{
    // Scanner uses other queue to send events. We must edit UI in the main queue
    dispatch_async(dispatch_get_main_queue(), ^{
        // Decode the characteristic data
        NSData *data = characteristic.value;
        uint8_t *array = (uint8_t*) data.bytes;
        
        if ([characteristic.UUID isEqual:batteryLevelCharacteristicUUID])
        {
            UInt8 batteryLevel = [CharacteristicReader readUInt8Value:&array];
            NSString* text = [[NSString alloc] initWithFormat:@"%d%%", batteryLevel];
            [battery setTitle:text forState:UIControlStateDisabled];
            
            if (battery.tag == 0)
            {
                // If battery level notifications are available, enable them
                if (([characteristic properties] & CBCharacteristicPropertyNotify) > 0)
                {
                    battery.tag = 1; // mark that we have enabled notifications
                    
                    // Enable notification on data characteristic
                    [peripheral setNotifyValue:YES forCharacteristic:characteristic];
                }
            }
        }
        else if ([characteristic.UUID isEqual:bpmBloodPressureMeasurementCharacteristicUUID] ||
                 [characteristic.UUID isEqual:bpmIntermediateCuffPressureCharacteristicUUID])
        {
            UInt8 flags = [CharacteristicReader readUInt8Value:&array];
            BOOL kPa = (flags & 0x01) > 0;
            BOOL timestampPresent = (flags & 0x02) > 0;
            BOOL pulseRatePresent = (flags & 0x04) > 0;
            
            // Update units
            if (kPa)
            {
                self.systolicUnit.text = @"kPa";
                self.diastolicUnit.text = @"kPa";
                self.meanApUnit.text = @"kPa";
            }
            else
            {
                self.systolicUnit.text = @"mmHg";
                self.diastolicUnit.text = @"mmHg";
                self.meanApUnit.text = @"mmHg";
            }
            
            // Read main values
            if ([characteristic.UUID isEqual:bpmBloodPressureMeasurementCharacteristicUUID])
            {
                float systolicValue = [CharacteristicReader readSFloatValue:&array];
                float diastolicValue = [CharacteristicReader readSFloatValue:&array];
                float meanApValue = [CharacteristicReader readSFloatValue:&array];
                
                self.systolic.text = [NSString stringWithFormat:@"%.1f", systolicValue];
                self.diastolic.text = [NSString stringWithFormat:@"%.1f", diastolicValue];
                self.meanAp.text = [NSString stringWithFormat:@"%.1f", meanApValue];
                
                self.systolicUnit.hidden = NO;
                self.diastolicUnit.hidden = NO;
                self.meanApUnit.hidden = NO;
            }
            else
            {
                float systolicValue = [CharacteristicReader readSFloatValue:&array];
                array += 4;
                
                self.systolic.text = [NSString stringWithFormat:@"%.1f", systolicValue];
                self.diastolic.text = @"n/a";
                self.meanAp.text = @"n/a";
                
                self.systolicUnit.hidden = NO;
                self.diastolicUnit.hidden = YES;
                self.meanApUnit.hidden = YES;
            }
            
            // Read timestamp
            if (timestampPresent)
            {
                NSDate* date = [CharacteristicReader readDateTime:&array];
                
                NSDateFormatter *dateFormat = [[NSDateFormatter alloc] init];
                [dateFormat setDateFormat:@"dd.MM.yyyy, hh:mm"];
                NSString* dateFormattedString = [dateFormat stringFromDate:date];
                
                self.timestamp.text = dateFormattedString;
            }
            else
            {
                self.timestamp.text = @"n/a";
            }
            
            // Read pulse
            if (pulseRatePresent)
            {
                float pulseValue = [CharacteristicReader readSFloatValue:&array];
                self.pulse.text = [NSString stringWithFormat:@"%.1f", pulseValue];
            }
            else
            {
                self.pulse.text = @"-";
            }
        }
    });
}

@end

File: nRF Toolbox/BPM/BPMViewController.swift
//
//  BMPViewController.swift
//  nRF Toolbox
//
//  Created by Mostafa Berg on 06/05/16.
//  Copyright © 2016 Nordic Semiconductor. All rights reserved.
//

import UIKit
import CoreBluetooth

class BPMViewController: BaseViewController, CBCentralManagerDelegate, CBPeripheralDelegate, ScannerDelegate, StoryboardInstantiable {
    
    //MARK: - ViewController Properties
    var bpmServiceUUID                                : CBUUID
    var bpmBloodPressureMeasurementCharacteristicUUID : CBUUID
    var bpmIntermediateCuffPressureCharacteristicUUID : CBUUID
    var batteryServiceUUID                            : CBUUID
    var batteryLevelCharacteristicUUID                : CBUUID
    var bluetoothManager                              : CBCentralManager?
    var connectedPeripheral                           : CBPeripheral?
    
    //MARK: - Referencing Outlets
    @IBOutlet weak var deviceName: UILabel!
    @IBOutlet weak var battery: UIButton!
    @IBOutlet weak var verticalLabel: UILabel!
    @IBOutlet weak var connectionButton: UIButton!
    @IBOutlet weak var systolic: UILabel!
    @IBOutlet weak var systolicUnit: UILabel!
    @IBOutlet weak var diastolic: UILabel!
    @IBOutlet weak var diastolicUnit: UILabel!
    @IBOutlet weak var meanAp: UILabel!
    @IBOutlet weak var meanApUnit: UILabel!
    @IBOutlet weak var pulse: UILabel!
    @IBOutlet weak var timestamp: UILabel!
    
    //MARK: - Referencing Actions
    @IBAction func connectionButtonTapped(_ sender: AnyObject) {
        if connectedPeripheral != nil {
            bluetoothManager?.cancelPeripheralConnection(connectedPeripheral!)
        }
    }
    @IBAction func aboutButtonTapped(_ sender: AnyObject) {
        self.showAbout(message: AppUtilities.getHelpTextForService(service: .bpm))
    }

    //MARK: - UIViewController methods
    required init?(coder aDecoder: NSCoder) {
        bpmServiceUUID                                = CBUUID(string: ServiceIdentifiers.bpmServiceUUIDString)
        bpmBloodPressureMeasurementCharacteristicUUID = CBUUID(string: ServiceIdentifiers.bpmBloodPressureMeasurementCharacteristicUUIDString)
        bpmIntermediateCuffPressureCharacteristicUUID = CBUUID(string: ServiceIdentifiers.bpmIntermediateCuffPressureCharacteristicUUIDString)
        batteryServiceUUID                            = CBUUID(string: ServiceIdentifiers.batteryServiceUUIDString)
        batteryLevelCharacteristicUUID                = CBUUID(string: ServiceIdentifiers.batteryLevelCharacteristicUUIDString)
        super.init(coder: aDecoder)
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.verticalLabel.transform = CGAffineTransform(translationX: -(verticalLabel.frame.width/2) + (verticalLabel.frame.height / 2), y: 0.0).rotated(by: -.pi / 2);
    }
    
    //MARK: - BPMViewController Implementation
    
    @objc func didEnterBackgroundCallback(notification aNotification: Notification) {
        let name = connectedPeripheral?.name ?? "peripheral"
        AppUtilities.showBackgroundNotification(message: "You are still connected to \(name). It will collect data also in background.")
    }
    
    @objc func didBecomeActiveCallback(notification aNotification: Notification) {
        UIApplication.shared.cancelAllLocalNotifications()
    }
    
    func clearUI() {
        deviceName.text = "DEFAULT BPM"
        battery.tag = 0
        battery.setTitle("n/a", for: .disabled)
        
        systolicUnit.isHidden     = true
        diastolicUnit.isHidden    = true
        meanApUnit.isHidden       = true
        systolic.text           = "-"
        diastolic.text          = "-"
        meanAp.text             = "-"
        pulse.text              = "-"
        timestamp.text          = "-"
    }

    
    //MARK: - ScannerDelegate methods
    
    func centralManagerDidSelectPeripheral(withManager aManager: CBCentralManager, andPeripheral aPeripheral: CBPeripheral) {
        clearUI()
        connectedPeripheral = aPeripheral
        connectedPeripheral?.delegate = self
        bluetoothManager = aManager
        bluetoothManager?.delegate = self

        let connectionOptions = NSDictionary(object: NSNumber(value: true as Bool), forKey: CBConnectPeripheralOptionNotifyOnNotificationKey as NSCopying)
        bluetoothManager?.connect(aPeripheral, options: connectionOptions as? [String : AnyObject])
    }
    
    //MARK: - CBCentralManagerDelegate
    
    func centralManagerDidUpdateState(_ central: CBCentralManager) {
        if central.state == .poweredOff {
            print("Bluetooth powered off")
        } else {
            print("Bluetooth powered on")
        }
    }
    
    func centralManager(_ central: CBCentralManager, didConnect peripheral: CBPeripheral) {
        DispatchQueue.main.async {
            self.deviceName.text = peripheral.name
            self.connectionButton.setTitle("DISCONNECT", for: .normal)
        
            //Following if condition display user permission alert for background notification
            if UIApplication.instancesRespond(to: #selector(UIApplication.registerUserNotificationSettings(_:))) {
                UIApplication.shared.registerUserNotificationSettings(UIUserNotificationSettings(types: [.alert, .sound], categories: nil))
            }
            NotificationCenter.default.addObserver(self, selector: #selector(self.didEnterBackgroundCallback(notification:)), name: UIApplication.didEnterBackgroundNotification, object: nil)
            NotificationCenter.default.addObserver(self, selector: #selector(self.didBecomeActiveCallback(notification:)), name: UIApplication.didBecomeActiveNotification, object: nil)
        }
        
        connectedPeripheral = peripheral
        peripheral.discoverServices([bpmServiceUUID, batteryServiceUUID])
    }
    
    func centralManager(_ central: CBCentralManager, didFailToConnect peripheral: CBPeripheral, error: Error?) {
        // Scanner uses other queue to send events. We must edit UI in the main queue
        DispatchQueue.main.async {
            AppUtilities.showAlert(title: "Error", andMessage: "Connecting to peripheral failed. Try again", from: self)
            self.connectionButton.setTitle("CONNECT", for: .normal)
            self.connectedPeripheral = nil
            self.clearUI()
        }
    }
    
    func centralManager(_ central: CBCentralManager, didDisconnectPeripheral peripheral: CBPeripheral, error: Error?) {
        // Scanner uses other queue to send events. We must edit UI in the main queue
        DispatchQueue.main.async {
            self.connectionButton.setTitle("CONNECT", for: .normal)
            self.connectedPeripheral = nil
            
            if AppUtilities.isApplicationInactive() {
                let name = peripheral.name ?? "Peripheral"
                AppUtilities.showBackgroundNotification(message: "\(name) is disconnected.")
            }
            
            NotificationCenter.default.removeObserver(self, name: UIApplication.didBecomeActiveNotification, object: nil)
            NotificationCenter.default.removeObserver(self, name: UIApplication.didEnterBackgroundNotification, object: nil)
        }
    }

    //MARK: - CBPeripheralDelegate
    func peripheral(_ peripheral: CBPeripheral, didDiscoverServices error: Error?) {
        guard error == nil else {
            print("An error occured while discovering services: \(error!.localizedDescription)")
            bluetoothManager!.cancelPeripheralConnection(peripheral)
            return
        }
        
        for aService : CBService in peripheral.services! {
            if aService.uuid == batteryServiceUUID {
                peripheral.discoverCharacteristics([batteryLevelCharacteristicUUID], for: aService)
            } else if aService.uuid == bpmServiceUUID {
                peripheral.discoverCharacteristics(
                    [bpmBloodPressureMeasurementCharacteristicUUID,bpmIntermediateCuffPressureCharacteristicUUID],
                    for: aService)
            }
        }
    }
    
    func peripheral(_ peripheral: CBPeripheral, didDiscoverCharacteristicsFor service: CBService, error: Error?) {
        guard error == nil else {
            print("Error occurred while discovering characteristic: \(error!.localizedDescription)")
            bluetoothManager!.cancelPeripheralConnection(peripheral)
            return
        }
        
        if service.uuid == bpmServiceUUID {
            for aCharacteristic: CBCharacteristic in service.characteristics! {
                if aCharacteristic.uuid == bpmBloodPressureMeasurementCharacteristicUUID ||
                    aCharacteristic.uuid == bpmIntermediateCuffPressureCharacteristicUUID {
                    peripheral.setNotifyValue(true, for: aCharacteristic)
                }
            }
        } else if service.uuid == batteryServiceUUID {
            for aCharacteristic: CBCharacteristic in service.characteristics! {
                if aCharacteristic.uuid == batteryLevelCharacteristicUUID {
                    peripheral.readValue(for: aCharacteristic)
                    break
                }
            }
        }
    }
    
    func peripheral(_ peripheral: CBPeripheral, didUpdateValueFor characteristic: CBCharacteristic, error: Error?) {
        guard error == nil else {
            print("Error occurred while updating characteristic value: \(error!.localizedDescription)")
            return
        }
        
        // Scanner uses other queue to send events. We must edit UI in the main queue
        DispatchQueue.main.async {
            if characteristic.uuid == self.batteryLevelCharacteristicUUID {
                // Decode the characteristic data
                let data = characteristic.value;
                var pointer = UnsafeMutablePointer<UInt8>(mutating: (data! as NSData).bytes.bindMemory(to: UInt8.self, capacity: data!.count))
                let batteryLevel = CharacteristicReader.readUInt8Value(ptr: &pointer)
                let text = "\(batteryLevel)%"
                self.battery.setTitle(text, for: .disabled)
                
                if self.battery.tag == 0 {
                    if characteristic.properties.rawValue & CBCharacteristicProperties.notify.rawValue > 0 {
                        peripheral.setNotifyValue(true, for: characteristic)
                        self.battery.tag = 1
                    }
                }
            } else if characteristic.uuid == self.bpmBloodPressureMeasurementCharacteristicUUID ||
                characteristic.uuid == self.bpmIntermediateCuffPressureCharacteristicUUID {
                let data = characteristic.value
                var pointer = UnsafeMutablePointer<UInt8>(mutating: (data! as NSData).bytes.bindMemory(to: UInt8.self, capacity: data!.count))
                let flags = CharacteristicReader.readUInt8Value(ptr: &pointer)
                let kPA              : Bool = (flags & 0x01) > 0
                let timestampPresent : Bool = (flags & 0x02) > 0
                let pulseRatePresent : Bool = (flags & 0x04) > 0
                
                if kPA == true {
                    self.systolicUnit.text  = "kPa"
                    self.diastolicUnit.text = "kPa"
                    self.meanApUnit.text    = "kPa"
                } else {
                    self.systolicUnit.text  = "mmHg"
                    self.diastolicUnit.text = "mmHg"
                    self.meanApUnit.text    = "mmHg"
                }
                
                //Read main values
                if characteristic.uuid == self.bpmBloodPressureMeasurementCharacteristicUUID {
                    let systolicValue  = CharacteristicReader.readSFloatValue(ptr: &pointer)
                    let diastolicValue = CharacteristicReader.readSFloatValue(ptr: &pointer)
                    let meanApValue    = CharacteristicReader.readSFloatValue(ptr: &pointer)
                    
                    self.systolic.text = String(format: "%.1f", systolicValue)
                    self.diastolic.text = String(format: "%.1f", diastolicValue)
                    self.meanAp.text = String(format: "%.1f", meanApValue)
                    
                    self.systolicUnit.isHidden    = false
                    self.diastolicUnit.isHidden   = false
                    self.meanApUnit.isHidden      = false
                } else {
                    let systolicValue = CharacteristicReader.readSFloatValue(ptr: &pointer)
                    pointer += 4
                    
                    self.systolic.text = String(format: "%.1f", systolicValue)
                    self.diastolic.text = "n/a"
                    self.meanAp.text = "n/a"
                    
                    self.systolicUnit.isHidden    = false
                    self.diastolicUnit.isHidden   = true
                    self.meanApUnit.isHidden      = true
                }
                
                // Read timestamp
                if timestampPresent {
                    let date = CharacteristicReader.readDateTime(ptr: &pointer)
                    let dateFormatter = DateFormatter()
                    dateFormatter.dateFormat = "dd.MM.yyy, hh:mm"
                    let dateformattedString = dateFormatter.string(from: date)
                    self.timestamp.text = dateformattedString
                } else {
                    self.timestamp.text = "n/a"
                }

                // Read pulse
                if pulseRatePresent {
                    let pulseValue = CharacteristicReader.readSFloatValue(ptr: &pointer)
                    self.pulse.text = String(format: "%.1f", pulseValue)
                } else {
                    self.pulse.text = "-"
                }
            }
        }
    }

    //MARK: - Segue handling
    override func shouldPerformSegue(withIdentifier identifier: String, sender: Any?) -> Bool {
        return identifier != "scan" || connectedPeripheral == nil
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if segue.identifier == "scan" {
            let nc = segue.destination as! UINavigationController
            let controller = nc.children.first as! ScannerViewController
            controller.filterUUID = bpmServiceUUID
            controller.delegate = self
        }
    }
}
